staging/src/k8s.io/component-base/zpages/statusz/statusz.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statusz
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"net/http"
+	"strings"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	funcMap = template.FuncMap{
+		"ToLower": strings.ToLower,
+	}
+	headerTmpl *template.Template
+	dataTmpl   *template.Template
+	reg        statuszRegistry = registry{}
+)
+
+const (
+	headerTemplate = `
+------------------------------------------------------------------------
+title: {{.ComponentName}} statusz
+description: details of the status data that {{.ComponentName}} reports.
+warning: This endpoint is not meant to be machine parseable and is for debugging purposes only.
+------------------------------------------------------------------------
+`
+
+	dataTemplate = `
+Started: {{.StartTime}}
+Up: {{.Uptime}}
+Go version: {{.GoVersion}}
+Binary version: {{.BinaryVersion}}
+Emulation version: {{.EmulationVersion}}
+Minimum Compatibility version: {{.MinCompatibilityVersion}}
+
+List of useful endpoints
+--------------
+{{- range $name, $link := .UsefulLinks}}
+{{$name}}:{{$link -}}
+{{- end }}
+`
+)
+
+type headerFields struct {
+	ComponentName string
+}
+
+type contentFields struct {
+	StartTime               string
+	Uptime                  string
+	GoVersion               string
+	BinaryVersion           string
+	EmulationVersion        string
+	MinCompatibilityVersion string
+	UsefulLinks             map[string]string
+}
+
+type mux interface {
+	Handle(path string, handler http.Handler)
+}
+
+func Install(m mux, componentName string) {
+	err := initializeTemplates()
+	if err != nil {
+		klog.Errorf("error while parsing gotemplates: %v", err)
+		return
+	}
+	m.Handle("/statusz", handleStatusz(componentName, headerTmpl, dataTmpl))
```

Comment:
We may want to make the subpaths entail components to allow querying them individually as well. The older healthchecks allow doing so for the constructs that affect them directly (`/healthz/poststarthook/start-apiserver-admission-initializer` or `/healthz/etcd`, for instance).

Reply:
WRT https://github.com/kubernetes/enhancements/pull/4830#discussion_r1790764661, would it make sense to supply the version (and format) as a parameter instead, to allow for component-subpaths (e.g., `/statusz/apiserver?format=json&v=2`)? OR make component subpaths require version at all times, like so: `/statusz/v2/apiserver?format=json`?

Reply:
Overall, I believe the final path may resemble: `/statusz/vX/apiserver` while the format is negotiated through the header.

Reply:
isnt that redundant? Every component runs on a a different port, so doing a request like 'http://localhost:<port>/statusz' would show statusz for that specific component. Also the header of the plain text response being returned contains the component name too.

Reply:
> Every component runs on a a different port, so doing a request like 'http://localhost/:/statusz' would show statusz for that specific component.

So, that would require the user to _know_ the component ports beforehand. This is possibly why the existing healthchecks offer a generalized way to query that information by specifying the component name without needing a port number. The multiplexer should resolve names to ports IMHO for a much better DX.

> Also the header of the plain text response being returned contains the component name too.

Since Prometheus negotiations concern only the `content-type` here, the component name maybe extraneous as it's already being specified in the query and the returned payload.

Reply:
> So, that would require the user to know the component ports beforehand

That's true. Does the "specifying component in the url as subpath" support exist for other endpoints other than healthz? It does seem weird for some endpoints to have that support and some not. Also curious for @liggitt 's thoughts on this, since you have already taken a pass at this PR once

Reply:
> Does the "specifying component in the url as subpath" support exist for other endpoints other than healthz?

I think so, all existing healthchecks at the moment support the same set of endpoints (you can check those by appending the `?verbose` param) between them.

That being said, `/flagz` and `/statusz` are conceptually different from the existing set of healthchecks in the sense that they do not introduce their own signals but rather relay certain stats from component endpoints (instead of `poststarthook`s and `etcd`, like the existing ones do), which are something that they both should share, based on convention.